frontend-react/src/shared/Tooltip.tsx

```diff
@@ -0,0 +1,215 @@
+import {
+    Placement,
+    useFloating,
+    autoUpdate,
+    flip,
+    shift,
+    useHover,
+    useFocus,
+    useDismiss,
+    useRole,
+    useInteractions,
+    useMergeRefs,
+    FloatingPortal,
+    offset,
+    OffsetOptions,
+} from "@floating-ui/react";
+import classNames from "classnames";
+import React from "react";
+
+export const TOOLTIP_POSITIONS = ["top", "bottom", "right", "left"];
+
+export interface TooltipOptions {
+    initialOpen?: boolean;
+    placement?: Placement;
+    open?: boolean;
+    onOpenChange?: (open: boolean) => void;
+    offsetBy?: OffsetOptions;
+}
+
+export function useTooltip({
+    initialOpen = false,
```

Comment:
Minor, minor point, but I wonder if we could just infer that this is controlled/uncontrolled.  Like if `open` is a boolean, it's controlled; otherwise, it's uncontrolled?